File: src/main/java/eladkay/quaritum/common/block/base/BlockModFlower.java
package eladkay.quaritum.common.block.base;

import eladkay.quaritum.api.animus.IFlower;
import eladkay.quaritum.client.core.ModelHandler;
import eladkay.quaritum.common.Quaritum;
import eladkay.quaritum.common.core.CreativeTab;
import net.minecraft.block.Block;
import net.minecraft.block.BlockBush;
import net.minecraft.block.material.Material;
import net.minecraft.block.properties.IProperty;
import net.minecraft.client.renderer.ItemMeshDefinition;
import net.minecraft.item.EnumRarity;
import net.minecraft.item.ItemStack;
import net.minecraftforge.fml.common.registry.GameRegistry;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

/**
 * @author WireSegal
 *         Created at 4:18 PM on 4/30/16.
 */
public abstract class BlockModFlower extends BlockBush implements IFlower, ModelHandler.IModBlock {

    private String[] variants;
    private String bareName;

    public BlockModFlower(String name, Material materialIn, String... variants) {
        super(materialIn);
        this.variants = variants;
        if (variants.length == 0) {
            this.variants = new String[] {name};
        }
        this.bareName = name;
        this.setUnlocalizedName(name);

        if (shouldRegisterInCreative())
            CreativeTab.set(this);
    }

    public boolean shouldRegisterInCreative() {
        return true;
    }

    public boolean shouldHaveItemForm() {
        return true;
    }

    @Override
    public Block setUnlocalizedName(String name) {
        super.setUnlocalizedName(name);
        setRegistryName(name);
        GameRegistry.register(this);
        if (shouldHaveItemForm())
            GameRegistry.register(new ItemModBlock(this)/*, new ResourceLocation(LibMisc.MOD_ID, name)*/);
        else
            Quaritum.proxy.addToVariantCache(this);
        //ModelHandler.variantCache.add(this);

        return this;
    }

    @Nonnull
    @Override
    public String getBareName() {
        return bareName;
    }

    @Nonnull
    @Override
    public String[] getVariants() {
        return variants;
    }

    @Nullable
    @Override
    public Class<Enum> getVariantEnum() {
        return null;
    }

    @Nonnull
    @Override
    public IProperty[] getIgnoredProperties() {
        return new IProperty[0];
    }

    @Override

    public ItemMeshDefinition getCustomMeshDefinition() {
        return null;
    }

    @Nonnull
    @Override
    public EnumRarity getBlockRarity(ItemStack stack) {
        return EnumRarity.COMMON;
    }
}

File: src/main/java/eladkay/quaritum/common/block/base/BlockModFlower.kt
package eladkay.quaritum.common.block.base

import com.teamwizardry.librarianlib.common.base.block.IModBlock
import com.teamwizardry.librarianlib.common.base.block.ItemModBlock
import com.teamwizardry.librarianlib.common.util.VariantHelper
import net.minecraft.block.Block
import net.minecraft.block.BlockBush
import net.minecraft.block.material.MapColor
import net.minecraft.block.material.Material
import net.minecraft.item.ItemBlock
import net.minecraftforge.fml.common.Loader

/**
 * @author WireSegal
 * *         Created at 4:18 PM on 4/30/16.
 */
abstract class BlockModFlower(name: String, materialIn: Material, color: MapColor, vararg variants: String) : BlockBush(materialIn, color), IModBlock {

    constructor(name: String, materialIn: Material, vararg variants: String) : this(name, materialIn, materialIn.materialMapColor, *variants)

    final override val variants: Array<out String>

    override val bareName: String = name
    val modId: String

    val itemForm: ItemBlock? by lazy { createItemForm() }

    init {
        modId = Loader.instance().activeModContainer().modId
        this.variants = VariantHelper.beginSetupBlock(name, variants)
        @Suppress("LeakingThis")
        VariantHelper.finishSetupBlock(this, name, itemForm, creativeTab)
    }

    override fun setUnlocalizedName(name: String): Block {
        super.setUnlocalizedName(name)
        VariantHelper.setUnlocalizedNameForBlock(this, modId, name, itemForm)
        return this
    }

    open fun createItemForm(): ItemBlock? {
        return ItemModBlock(this)
    }
}
